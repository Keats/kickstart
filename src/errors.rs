use std::error::Error as StdError;
use std::fmt;
use std::io;
use std::path::PathBuf;
use std::result;

use tera;
use toml;

/// A crate private constructor for `Error`.
pub(crate) fn new_error(kind: ErrorKind) -> Error {
    Error { kind, source: None }
}

/// A type alias for `Result<T, kickstart::Error>`.
pub type Result<T> = result::Result<T, Error>;

/// The Error type
#[derive(Debug)]
pub struct Error {
    /// Kind of error
    pub kind: ErrorKind,
    pub source: Option<Box<dyn StdError>>,
}

/// The specific type of an error.
#[derive(Debug)]
pub enum ErrorKind {
    MissingTemplateDefinition,
    InvalidInput {
        msg: String,
    },
    InvalidTemplate,
    UnreadableStdin,
    /// An error while cloning a repository
    Git {
        err: io::Error,
    },
    /// An error while doing IO (reading/writing files)
    Io {
        err: io::Error,
        path: PathBuf,
    },
    /// An error when rendering a template, where a template can also refer to a filename
    /// in the case kickstart
    Tera {
        err: tera::Error,
        path: Option<PathBuf>,
    },
    /// An error while deserializing a template.toml into a struct
    Toml {
        err: toml::de::Error,
    },
    /// Hints that destructuring should not be exhaustive.
    ///
    /// This enum may grow additional variants, so this makes sure clients
    /// don't count on exhaustive matching. (Otherwise, adding a new variant
    /// could break existing code.)
    #[doc(hidden)]
    __Nonexhaustive,
}

impl From<io::Error> for Error {
    fn from(err: io::Error) -> Error {
        new_error(ErrorKind::Io { err, path: PathBuf::new() })
    }
}

impl StdError for Error {
    fn description(&self) -> &str {
        match self.kind {
            ErrorKind::Io { ref err, .. } => err.description(),
            ErrorKind::Tera { ref err, .. } => err.description(),
            ErrorKind::InvalidInput { ref msg, .. } => msg,
            ErrorKind::InvalidTemplate => "invalid template",
            ErrorKind::MissingTemplateDefinition => "missing template.toml",
            ErrorKind::UnreadableStdin => "couldn't read from stdin",
            ErrorKind::Git { ref err } => err.description(),
            ErrorKind::Toml { ref err } => err.description(),
            _ => unreachable!(),
        }
    }

    fn cause(&self) -> Option<&dyn StdError> {
        match self.kind {
            ErrorKind::Io { ref err, .. } => Some(err),
            ErrorKind::Tera { ref err, .. } => Some(err),
            ErrorKind::Toml { ref err } => Some(err),
            ErrorKind::Git { ref err } => Some(err),
            ErrorKind::InvalidInput { ref msg, .. } => None,
            ErrorKind::InvalidTemplate => None,
            ErrorKind::MissingTemplateDefinition => None,
            ErrorKind::UnreadableStdin => None,
            _ => unreachable!(),
        }
    }

    fn source(&self) -> Option<&(dyn StdError + 'static)> {
        let mut source = self.source.as_ref().map(|c| &**c);
        if source.is_none() {
            if let ErrorKind::Tera { ref err, .. } = self.kind {
                source = err.source();
            }
        }

        source
    }
}

impl fmt::Display for Error {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self.kind {
            ErrorKind::Io { ref err, ref path } => write!(f, "{}: {:?}", err, path),
            ErrorKind::Tera { ref err, ref path } => {
                if let Some(p) = path {
                    write!(f, "{}: {:?}", err, p)
                } else {
                    write!(f, "{}: rendering a one-off template", err)
                }
            }
            ErrorKind::Git { ref err } => write!(f, "Could not clone the repository: {}", err),
            ErrorKind::Toml { ref err } => write!(f, "Invalid TOML: {}", err),
            ErrorKind::MissingTemplateDefinition => write!(f, "The template.toml is missing"),
            ErrorKind::UnreadableStdin => write!(f, "Unable to read from stdin"),
            ErrorKind::InvalidInput { ref msg } => write!(f, "{}", msg),
            ErrorKind::InvalidTemplate => write!(f, "The template.toml is invalid"),
            _ => unreachable!(),
        }
    }
}
