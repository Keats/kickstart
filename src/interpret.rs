use regex::Regex;

use crate::errors::{new_error, ErrorKind, Result};

/// Interpret a yes/no question to the user
pub fn interpret_bool(input: &str, default: bool) -> Result<bool> {
    match &*input {
        "y" | "Y" | "yes" | "YES" | "true" => Ok(true),
        "n" | "N" | "no" | "NO" | "false" => Ok(false),
        "" => Ok(default),
        _ => Err(new_error(ErrorKind::InvalidInput {
            msg: format!("Invalid choice: '{}'\n", input),
        })),
    }
}

/// Interpret a question to the user where they can write any string
pub fn interpret_string(input: &str, default: &str, validation: &Option<String>) -> Result<String> {
    match &*input {
        "" => Ok(default.to_string()),
        _ => {
            if let Some(ref pattern) = validation {
                let re = Regex::new(pattern).unwrap();
                if re.is_match(&input) {
                    Ok(String::from(input))
                } else {
                    Err(new_error(ErrorKind::InvalidInput {
                        msg: format!("The value needs to pass the regex: {}\n", pattern),
                    }))
                }
            } else {
                Ok(String::from(input))
            }
        }
    }
}

/// Interpret a question to the user where they can write an integer
pub fn interpret_integer(input: &str, default: i64) -> Result<i64> {
    match &*input {
        "" => Ok(default),
        _ => match input.parse::<i64>() {
            Ok(i) => Ok(i),
            Err(_) => Err(new_error(ErrorKind::InvalidInput {
                msg: format!("Invalid integer: '{}'\n", input),
            })),
        },
    }
}

/// Interpret a choice between various options
pub fn interpret_choices(
    input: &str,
    default: &toml::Value,
    choices: &[toml::Value],
) -> Result<toml::Value> {
    match &*input {
        "" => Ok(default.clone()),
        _ => {
            if let Ok(num) = input.parse::<usize>() {
                if num > choices.len() {
                    Err(new_error(ErrorKind::InvalidInput {
                        msg: format!("Invalid choice: '{}'\n", input),
                    }))
                } else {
                    Ok(choices[num - 1].clone())
                }
            } else {
                Err(new_error(ErrorKind::InvalidInput {
                    msg: format!("Invalid choice: '{}'\n", input),
                }))
            }
        }
    }
}
