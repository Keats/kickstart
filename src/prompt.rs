use std::io::{self, BufRead, Write};

use toml;

use crate::errors::{new_error, ErrorKind, Result};
use crate::interpret;
use crate::terminal;

/// Wait for user input and return what they typed
fn read_line() -> Result<String> {
    let stdin = io::stdin();
    let stdin = stdin.lock();
    let mut lines = stdin.lines();
    lines.next().and_then(|l| l.ok()).ok_or_else(|| new_error(ErrorKind::UnreadableStdin))
}

/// Ask a yes/no question to the user
pub fn ask_bool(prompt: &str, default: bool) -> Result<bool> {
    terminal::bool_question(prompt, default);
    let _ = io::stdout().flush();
    let input = read_line()?;

    let res = match interpret::interpret_bool(&input, default) {
        Ok(x) => x,
        Err(e) => {
            terminal::error(&format!("{}", e));
            ask_bool(prompt, default)?
        }
    };

    Ok(res)
}

/// Ask a question to the user where they can write any string
pub fn ask_string(prompt: &str, default: &str, validation: &Option<String>) -> Result<String> {
    terminal::basic_question(prompt, &default, validation);
    let _ = io::stdout().flush();
    let input = read_line()?;

    let res = match interpret::interpret_string(&input, default, validation) {
        Ok(x) => x,
        Err(e) => {
            terminal::error(&format!("{}", e));
            ask_string(prompt, default, validation)?
        }
    };

    Ok(res)
}

/// Ask a question to the user where they can write an integer
pub fn ask_integer(prompt: &str, default: i64) -> Result<i64> {
    terminal::basic_question(prompt, &default, &None);
    let _ = io::stdout().flush();
    let input = read_line()?;

    let res = match interpret::interpret_integer(&input, default) {
        Ok(x) => x,
        Err(e) => {
            terminal::error(&format!("{}", e));
            ask_integer(prompt, default)?
        }
    };

    Ok(res)
}

/// Ask users to make a choice between various options
pub fn ask_choices(
    prompt: &str,
    default: &toml::Value,
    choices: &[toml::Value],
) -> Result<toml::Value> {
    terminal::bold(&format!("{}: \n", prompt));
    let mut lines = vec![];
    let mut default_index = 1;

    for (index, choice) in choices.iter().enumerate() {
        terminal::bold(&format!("  {}. {}\n", index + 1, choice.as_str().unwrap()));

        lines.push(format!("{}", index + 1));
        if choice == default {
            default_index = index + 1;
        }
    }

    terminal::basic_question(
        &format!("  > Choose from {}..{}", 1, lines.len()),
        &default_index,
        &None,
    );

    let _ = io::stdout().flush();
    let input = read_line()?;

    let res = match interpret::interpret_choices(&input, default, choices) {
        Ok(x) => x,
        Err(e) => {
            terminal::error(&format!("{}", e));
            ask_choices(prompt, default, choices)?
        }
    };

    Ok(res)
}
