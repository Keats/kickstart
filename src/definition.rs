use std::collections::HashMap;

use serde::Deserialize;
use toml::Value;

use crate::interpret::{interpret_bool, interpret_choices, interpret_integer, interpret_string};
use crate::errors::{new_error, ErrorKind, Result};
use crate::prompt::{ask_bool, ask_choices, ask_integer, ask_string};

/// A condition for a question to be asked
#[derive(Debug, Clone, PartialEq, Deserialize)]
pub struct VariableCondition {
    pub name: String,
    pub value: Value,
}

/// A list of items to be deleted when `name` has `value`
#[derive(Debug, Clone, PartialEq, Deserialize)]
pub struct Cleanup {
    pub name: String,
    pub value: Value,
    pub paths: Vec<String>,
}

/// A question loaded from TOML
#[derive(Debug, Clone, PartialEq, Deserialize)]
pub struct Variable {
    /// The variable name in the final context
    pub name: String,
    /// A default value is required
    pub default: Value,
    /// The text asked to the user
    pub prompt: String,
    /// Only for questions with choices
    pub choices: Option<Vec<Value>>,
    /// A regex pattern to validate the input
    pub validation: Option<String>,
    /// Only ask this variable is the condition is true
    pub only_if: Option<VariableCondition>,
}

/// The full template struct we get fom loading a TOML file
#[derive(Debug, Clone, PartialEq, Deserialize)]
pub struct TemplateDefinition {
    /// Name of the template
    pub name: String,
    /// What this template is about
    pub description: Option<String>,
    /// Template version
    pub version: Option<String>,
    /// Version of the kickstart template spec
    pub kickstart_version: usize,
    /// Url of the template
    pub url: Option<String>,
    /// A list of the authors
    #[serde(default)]
    pub authors: Vec<String>,
    /// Some keywords/tags
    #[serde(default)]
    pub keywords: Vec<String>,
    /// The directory in which the template files are.
    /// Useful if a template has its own docs, README, CI and various files
    pub directory: Option<String>,
    /// Do not copy those directories/files
    #[serde(default)]
    pub ignore: Vec<String>,
    /// Conditionally delete some files/dirs based on generator values
    #[serde(default)]
    pub cleanup: Vec<Cleanup>,
    /// Do not pass those files through Tera
    /// http://cookiecutter.readthedocs.io/en/latest/advanced/copy_without_render.html
    #[serde(default)]
    pub copy_without_render: Vec<String>,
    /// All the questions for that template
    pub variables: Vec<Variable>,
}

pub enum Input {
    ArgumentOrDefault(HashMap<String, String>),
    Interactive,
}

impl TemplateDefinition {
    /// Ask all the questions of that template and return the answers.
    /// If `input` is `ArgumentOrDefault`, it will automatically pick the 
    /// argument if exists or default without prompting the user
    pub fn ask_questions(&self, input: Input) -> Result<HashMap<String, Value>> {
        // Tera context doesn't expose a way to get value from a context
        // so we store them in another hashmap
        let mut vals = HashMap::new();

        for var in &self.variables {
            // Skip the question if the value is different from the condition
            if let Some(ref cond) = var.only_if {
                if let Some(val) = vals.get(&cond.name) {
                    if *val != cond.value {
                        continue;
                    }
                } else {
                    // Not having it means we didn't even ask the question
                    continue;
                }
            }

            if let Some(ref choices) = var.choices {
                let res = if let Input::ArgumentOrDefault(argument) = &input {
                    if let Some(input) = argument.get(&var.name) {
                        interpret_choices(input, &var.default, choices)?
                    } else {
                        var.default.clone()
                    }
                } else {
                    ask_choices(&var.prompt, &var.default, choices)?
                };
                vals.insert(var.name.clone(), res);
                continue;
            }

            match &var.default {
                Value::Boolean(b) => {
                    let res = if let Input::ArgumentOrDefault(argument) = &input {
                        if let Some(input) = argument.get(&var.name) {
                            interpret_bool(input, *b)?
                        } else {
                            *b                            
                        }
                    } else { ask_bool(&var.prompt, *b)? };
                    vals.insert(var.name.clone(), Value::Boolean(res));
                    continue;
                }
                Value::String(s) => {
                    let res = if let Input::ArgumentOrDefault(argument) = &input {
                        if let Some(input) = argument.get(&var.name) {
                            interpret_string(input, s, &var.validation)?
                        } else {
                            s.clone()
                        }
                    } else {
                        ask_string(&var.prompt, &s, &var.validation)?
                    };
                    vals.insert(var.name.clone(), Value::String(res));
                    continue;
                }
                Value::Integer(i) => {
                    let res = if let Input::ArgumentOrDefault(argument) = &input { 
                        if let Some(input) = argument.get(&var.name) {
                            interpret_integer(input, *i)?
                        } else {
                            *i
                        }
                    } else { ask_integer(&var.prompt, *i)? };
                    vals.insert(var.name.clone(), Value::Integer(res));
                    continue;
                }
                _ => return Err(new_error(ErrorKind::InvalidTemplate)),
            }
        }

        Ok(vals)
    }
}

#[cfg(test)]
mod tests {
    use toml;

    use super::*;

    #[test]
    fn can_load_template_and_work_with_no_input() {
        let tpl: TemplateDefinition = toml::from_str(
            r#"
            name = "Test template"
            description = "A description"
            kickstart_version = 1

            [[variables]]
            name = "project_name"
            default = "My project"
            prompt = "What's the name of your project?"

            [[variables]]
            name = "database"
            default = "postgres"
            prompt = "Which database to use?"
            choices = ["postgres", "mysql"]

            [[variables]]
            name = "pg_version"
            prompt = "Which version of Postgres?"
            default = "10.4"
            choices = ["10.4", "9.3"]
            only_if = { name = "database", value = "postgres" }

        "#,
        )
        .unwrap();

        assert_eq!(tpl.variables.len(), 3);
        let res = tpl.ask_questions(Input::ArgumentOrDefault(HashMap::new()));
        assert!(res.is_ok());
    }

    #[test]
    fn only_if_questions_are_skipped_if_cond_invalid() {
        let tpl: TemplateDefinition = toml::from_str(
            r#"
            name = "Test template"
            description = "A description"
            kickstart_version = 1

            [[variables]]
            name = "project_name"
            default = "My project"
            prompt = "What's the name of your project?"

            [[variables]]
            name = "database"
            default = "postgres"
            prompt = "Which database to use?"
            choices = ["postgres", "mysql"]

            [[variables]]
            name = "pg_version"
            prompt = "Which version of Postgres?"
            default = "10.4"
            choices = ["10.4", "9.3"]
            only_if = { name = "database", value = "mysql" }

        "#,
        )
        .unwrap();

        assert_eq!(tpl.variables.len(), 3);
        let res = tpl.ask_questions(Input::ArgumentOrDefault(HashMap::new()));
        assert!(res.is_ok());
        let res = res.unwrap();
        assert!(!res.contains_key("pg_version"));
    }

    #[test]
    fn nested_only_if_questions_are_skipped_if_initial_cond_invalid() {
        let tpl: TemplateDefinition = toml::from_str(
            r#"
            name = "Test template"
            description = "A description"
            kickstart_version = 1

            [[variables]]
            name = "project_name"
            default = "My project"
            prompt = "What's the name of your project?"

            [[variables]]
            name = "database"
            default = "postgres"
            prompt = "Which database to use?"
            choices = ["postgres", "mysql"]

            [[variables]]
            name = "pg_version"
            prompt = "Which version of Postgres?"
            default = "10.4"
            choices = ["10.4", "9.3"]
            only_if = { name = "database", value = "mysql" }

            [[variables]]
            name = "pg_bouncer"
            prompt = "Add pgBouncer?"
            default = true
            only_if = { name = "pg_version", value = "10.4" }
        "#,
        )
        .unwrap();

        assert_eq!(tpl.variables.len(), 4);
        let res = tpl.ask_questions(Input::ArgumentOrDefault(HashMap::new()));
        assert!(res.is_ok());
        let res = res.unwrap();
        assert!(!res.contains_key("pg_version"));
        assert!(!res.contains_key("pg_bouncer"));
    }
}
