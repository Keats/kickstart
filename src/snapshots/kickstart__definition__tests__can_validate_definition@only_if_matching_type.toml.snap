---
source: src/definition.rs
expression: "&errs"
input_file: src/snapshots/validation/only_if_matching_type.toml
---
Ok(
    [
        "Variable `pg_version` depends on `database=true`, but the type of `database` is string",
    ],
)
